File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            ")": "(",
            "]": "[",
            "}": "{",
        }
        parens = []
        
        for char in s:
            if char in pairs:
                if len(parens) == 0 or pairs[char] != parens.pop():
                    return False
            else:
                parens.append(char)
        
        return len(parens) == 0

File: 20-valid-parentheses/20-valid-parentheses.go
import lls "github.com/emirpasic/gods/stacks/linkedliststack"

func isValid(s string) bool {
    parens := lls.New()
    pairs := map[rune]rune {
        ')': '(',
        ']': '[',
        '}': '{',
    }
    
    for _, r := range s {
        if r == ')' || r == ']' || r == '}' {
            if p, ok := parens.Pop(); !ok || p != pairs[r] {
                return false
            }
        } else {
            parens.Push(r)
        }
    }
    
    return parens.Empty()
}